[Tensor Parallel] Simplify distribute for MHA

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100046

This function is only called for nn.MHA or the custom MHA we use, and
if it is the former it is converted to the latter. So this check can actually
be an assert.

Differential Revision: [D45300396](https://our.internmc.facebook.com/intern/diff/D45300396/)

rohan-varma committed:
[Tensor Parallel] Simplify distribute for MHA

This function is only called for nn.MHA or the custom MHA we use, and
if it is the former it is converted to the latter. So this check can actually
be an assert.

Differential Revision: [D45300396](https://our.internmc.facebook.com/intern/diff/D45300396/)

[ghstack-poisoned]

awgu commented on torch/distributed/tensor/parallel/api.py:
```diff
@@ -334,24 +334,25 @@ def _parallelize_multihead_attn(
-                    _rowwise_parallelize_linear_fn,
-                    output_fn=parallel_style._prepare_output,  # type: ignore[arg-type, misc] # pyre-ignore[6]
-                )
+    assert isinstance(module, TensorParallelMultiheadAttention)
```
If we go for this change, let us add an assert message to provide more info?
```suggestion
    assert isinstance(module, TensorParallelMultiheadAttention), (
        f"Expects TensorParallelMultiheadAttention but got {type(module)}"
    )
```

rohan-varma committed:
Update on "[Tensor Parallel] Simplify distribute for MHA"


This function is only called for nn.MHA or the custom MHA we use, and
if it is the former it is converted to the latter. So this check can actually
be an assert.

Differential Revision: [D45300396](https://our.internmc.facebook.com/intern/diff/D45300396/)

[ghstack-poisoned]

wanchaol reviewed (approved):
thanks!

rohan-varma commented:
@pytorchbot merge -f "added label"
